hashdist/core/build_store.py

```diff
@@ -323,15 +349,80 @@ def delete(self, artifact_id):
     def _get_artifact_path(self, name, digest):
         return pjoin(self.artifact_root, name, digest[:SHORT_ARTIFACT_ID_LEN])
 
-    def resolve(self, artifact_id):
+    def _download_artifact(self, url,path):
+        # Provide a special case for local files
+        use_urllib = not SIMPLE_FILE_URL_RE.match(url)
+        if not use_urllib:
+            try:
+                stream = open(url[len('file:'):])
+            except IOError as e:
+                raise StoreNotFoundError(str(e))
+        else:
+            # Make request.
+            sys.stderr.write('Downloading %s...\n' % url)
+            try:
+                stream = urllib2.urlopen(url)
+            except urllib2.HTTPError, e:
+                msg = "urllib failed to download (code: %d): %s" % (e.code, url)
+                self.logger.warning(msg)
+                raise RemoteBuildStoreFetchError(msg)
+            except urllib2.URLError, e:
+                msg = "urllib failed to download (reason: %s): %s" % (e.reason, url)
+                self.logger.warning(msg)
+                raise RemoteBuildStoreFetchError(msg)
+
+        # Download file to a temporary file within self.packs_path, while hashing
+        # it.
+        self.logger.info("Downloading '%s'" % url)
+        temp_fd, temp_path = tempfile.mkstemp(prefix='downloading-', dir=self.artifact_root)
+        try:
+            f = os.fdopen(temp_fd, 'wb')
+            tee = HashingWriteStream(hashlib.sha256(), f)
+            if use_urllib:
+                if 'Content-Length' in stream.headers:
+                    progress = ProgressBar(int(stream.headers["Content-Length"]))
+                else:
+                    progress = ProgressSpinner()
+            try:
+                n = 0
+                while True:
+                    chunk = stream.read(self.chunk_size)
+                    if not chunk: break
+                    if use_urllib:
+                        n += len(chunk)
+                        progress.update(n)
+                    tee.write(chunk)
+            finally:
+                stream.close()
+                f.close()
+                if use_urllib:
+                    progress.finish()
+        except Exception as e:
+            # Remove temporary file if there was a failure
+            os.unlink(temp_path)
+            msg = "Unhandled Exception in Download: %s" % e
+            self.logger.warning(msg)
+            raise RemoteBuildStoreFetchError(msg)
+        os.chmod(temp_path, stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH)
+        os.system('cd %s; tar xzvf %s; rm -f %s' % (self.artifact_root,temp_path,temp_path))
```

Comment:
This will not raise an exception if there's a problem which might become very puzzling in a debug situation... so ideally replace with Python calls (and `subprocess.check_call` for the tar command). I think we have `safe_rmtree`  or similar that retries in case of network filesystems hickups and so on too?